File: problem-solving/algospot/fence/_solve.cpp
#include <cstdio>
#include <vector>

using namespace std;

int solve(vector<int> &v, int begin, int end);
int main(void)
{
	int c;
	scanf("%d", &c);

	while (c--)
	{
		int n;
		scanf("%d", &n);
		vector<int> v;
		for (int i = 0; i < n; ++i)
		{
			int temp;
			scanf("%d", &temp);
			v.push_back(temp);
		}
		printf("%d\n", solve(v, 0, v.size()));
	}
}
int findMidMax(vector<int> &v, int begin, int end);

int solve(vector<int> &v, int begin, int end)
{
	if (begin == end)
	{
		return 0;
	}
	if (begin + 1 == end)
	{
		return v[begin];
	}

	int mid = (begin + end) / 2;
	int leftMax = solve(v, begin, mid);
	int rightMax = solve(v, mid, end);
	int midMax = findMidMax(v, begin, end);

	return max(leftMax, max(rightMax, midMax));
}

int findMidMax(vector<int> &v, int begin, int end)
{
	int mid = (begin + end) / 2;
	int left = mid - 1;
	int right = mid;
	int minHeight = min(v[left], v[right]);
	int size = 2 * minHeight;
	while (begin < left || right + 1 < end)
	{
		if (begin < left && (right + 1 == end || v[left - 1] > v[right + 1]))
		{
			--left;
			minHeight = min(minHeight, v[left]);
		}
		else
		{
			++right;
			minHeight = min(minHeight, v[right]);
		}
		size = max(size, minHeight * (right - left + 1));
	}
	return size;
}

File: problem-solving/algospot/fence/solve.go
package main

import (
	"fmt"
)

func main() {
	var c int
	fmt.Scan(&c)

	for ; c > 0; c-- {
		var n int
		fmt.Scan(&n)
		arr := make([]int, n)
		for idx := range arr {
			fmt.Scan(&arr[idx])
		}
		fmt.Println(solve(arr))
	}
}

func solve(arr []int) int {
	if len(arr) == 0 {
		return 0
	}
	if len(arr) == 1 {
		return arr[0]
	}
	mid := len(arr) / 2
	leftMax := solve(arr[:mid])
	rightMax := solve(arr[mid:])
	midMax := findMidMax(arr)

	return max(leftMax, max(rightMax, midMax))
}

func findMidMax(arr []int) int {
	mid := len(arr) / 2
	left, right := mid-1, mid
	minHeight := min(arr[left], arr[right])
	ret := minHeight * 2
	for left > 0 || right+1 < len(arr) {
		if left > 0 && (right+1 == len(arr) || arr[left-1] > arr[right+1]) {
			left--
			minHeight = min(minHeight, arr[left])
		} else {
			right++
			minHeight = min(minHeight, arr[right])
		}
		ret = max(ret, minHeight*(right-left+1))
	}
	return ret
}

func min(a int, b int) int {
	if a > b {
		return b
	}
	return a
}
func max(a int, b int) int {
	if a < b {
		return b
	}
	return a
}
